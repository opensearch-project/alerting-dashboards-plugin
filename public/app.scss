/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

@import "./pages/MonitorDetails/components/MonitorOverview/MonitorOverview.scss";