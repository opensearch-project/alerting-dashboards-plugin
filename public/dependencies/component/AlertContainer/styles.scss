/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

.alert-container {
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  margin: 10px;
  overflow: hidden;

  .alert-container__scroll {
    display: flex;
    flex-grow: 1;
    flex-direction: column;
    overflow-y: auto;
  }

  &__flex-group {
    height: 100%;
  }

  &__modes {
    display: grid;
    grid-template-columns: 1fr 1fr;
    gap: 12px;
  }

  &__title-and-toggle {
    display: flex;
    justify-content: space-between;
  }

}
