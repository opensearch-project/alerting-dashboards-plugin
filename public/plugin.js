/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { PLUGIN_NAME } from '../utils/constants';

export class AlertingPlugin {
  constructor(initializerContext) {
    // can retrieve config from initializerContext
  }

  setup(core) {
    core.application.register({
      id: PLUGIN_NAME,
      title: 'Alerting',
      description: 'OpenSearch Dashboards Alerting Plugin',
      category: {
        id: 'opensearch',
        label: 'OpenSearch Plugins',
        order: 2000,
      },
      order: 4000,
      mount: async (params) => {
        const { renderApp } = await import('./app');
        const [coreStart] = await core.getStartServices();
        return renderApp(coreStart, params);
      },
    });
    return {};
  }

  start(core) {
    return {};
  }
}
