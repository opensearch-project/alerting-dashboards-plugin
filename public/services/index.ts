/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import NotificationService from './NotificationService';

export * from './services';
export { NotificationService };
