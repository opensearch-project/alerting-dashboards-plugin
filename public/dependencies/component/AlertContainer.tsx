/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useCallback, useEffect, useMemo, useState } from 'react';
import {
  EuiButton, EuiCallOut,
  EuiFlexGroup, EuiFlexItem,
  EuiSpacer, EuiText, EuiTitle,
} from '@elastic/eui';
import _ from 'lodash';
import { FieldArray, Formik } from 'formik';
import {
  getClient,
  getNotifications,
  getUISettings,
  NotificationService,
} from '../../services';
import {
  getInitialValues,
  getPlugins,
  submit,
} from '../../pages/CreateMonitor/containers/CreateMonitor/utils/helpers';
import DefineMonitor from '../../pages/CreateMonitor/containers/DefineMonitor';
import { formikToMonitor } from '../../pages/CreateMonitor/containers/CreateMonitor/utils/formikToMonitor';
import MonitorDetails from '../../pages/CreateMonitor/containers/MonitorDetails';
import ConfigureTriggers from '../../pages/CreateTrigger/containers/ConfigureTriggers';
import { unitToLabel } from '../../pages/CreateMonitor/components/Schedule/Frequencies/Interval';
import EnhancedAccordion from '../../components/FeatureAnywhereContextMenu/EnhancedAccordion';
import './styles.scss';

function AlertContainer(props: {content: string}) {
  const isDarkMode = getUISettings().get('theme:darkMode') || false;
  const setFlyout = () => null;
  const httpClient = getClient();
  const notifications = getNotifications();
  const searchType = '';
  const flyoutMode = 'olly';
  const history = {
    location: { pathname: '/create-monitor', search: props.content, hash: '', state: undefined },
    push: () => null,
    goBack: null,
  };
  const initialValues = useMemo(
    () =>
      getInitialValues({ ...history,  searchType }),
    []
  );
  const [isLoading, setIsLoading] = useState(false);
  const [isDisabled, setIsDisabled] = useState(false);
  const [plugins, setPlugins] = useState([]);
  const notificationService = useMemo(() => new NotificationService(httpClient), []);
  const [accordionsOpen, setAccordionsOpen] = useState<{[key: string]: boolean}>({ monitorDetails: false, defineMonitor: true});

  const onAccordionToggle = (key: string) => {
    if (key in accordionsOpen) {
      setAccordionsOpen({ ...accordionsOpen, [key]: !accordionsOpen[key]});
    }
  };

  const onCreate = useCallback( async (values, formikBag) => {
      submit({
        values,
        formikBag,
        history,
        updateMonitor: () => null,
        notifications,
        httpClient,
        onSuccess: async ({ monitor }) => {
          setIsDisabled(true)
          setIsLoading(false)
          notifications.toasts.addSuccess(`Monitor "${monitor.name}" successfully created.`);
        },
      });
  }, [history, notifications, httpClient]);
  const onSubmit = useCallback( async ({ handleSubmit, validateForm }) => {
    setIsLoading(true);
    const errors = await validateForm();

    if (Object.keys(errors).length > 0) {
      // Delay to allow errors to render
      setTimeout(() => {
        document
        .querySelector('.alert-container__scroll')
        ?.scrollTo({ top: 0, behavior: 'smooth' });
      }, 300);
      setIsLoading(false);
    }

    const hasErrorInMonitorDetails = 'name' in errors;
    const hasErrorInDefineMonitor = 'aggregations' in errors || 'groupBy' in errors ||
      'where' in errors || 'bucketValue' in errors;
    if (hasErrorInMonitorDetails != accordionsOpen.monitorDetails ||
      hasErrorInDefineMonitor != accordionsOpen.defineMonitor) {
      setAccordionsOpen({ ...accordionsOpen,
        monitorDetails: hasErrorInMonitorDetails || accordionsOpen.monitorDetails,
        defineMonitor: hasErrorInDefineMonitor || accordionsOpen.defineMonitor});
    }

    handleSubmit();
  }, [setIsLoading, accordionsOpen, setAccordionsOpen]);

  useEffect(() => {
    const updatePlugins = async () => {
      const newPlugins = await getPlugins(httpClient);
      setPlugins(newPlugins);
    };

    updatePlugins();
  }, []);

  return (
      <Formik initialValues={initialValues} onSubmit={onCreate} validateOnChange={false} >
        {(formikProps) => {
          const {values, handleSubmit, errors, touched, validateForm} = formikProps;
          const errorList = Object.entries(errors) as [string, string][];

          return (
            <div className="alert-container">
              {errorList.length > 0 && (
                <>
                  <EuiSpacer size="m" />
                  <EuiCallOut title="Address the following errors:" color="danger" iconType="alert">
                    <ul>
                      {
                        errorList.map(([k, v]) => (<li key={k}>{v}</li>))
                      }
                    </ul>
                  </EuiCallOut>
                </>
              )}
              <EuiSpacer size="m" />
              <div className="alert-container__scroll">
                <EnhancedAccordion
                  {...{
                    id: 'monitorDetails',
                    isOpen: accordionsOpen.monitorDetails,
                    onToggle: () => onAccordionToggle('monitorDetails'),
                    title: values.name,
                    subTitle: (
                      <>
                        <EuiText size="m" className="create-monitor__frequency">
                          <p>
                            Runs every {values.period.interval} <span>{unitToLabel[values.period.unit]}</span>
                          </p>
                        </EuiText>
                      </>
                    ),
                  }}
                >
                  <MonitorDetails
                    values={values}
                    errors={errors}
                    history={history}
                    httpClient={httpClient}
                    plugins={plugins}
                    isAd={false}
                    flyoutMode={flyoutMode} />
                </EnhancedAccordion>
                <EuiSpacer size="m" />

                <EnhancedAccordion
                  {...{
                    id: 'defineMonitor',
                    isOpen: accordionsOpen.defineMonitor,
                    onToggle: () => onAccordionToggle('defineMonitor'),
                    title: 'Advanced data source configuration',
                  }}
                >
                  <DefineMonitor
                    values={values}
                    errors={errors}
                    touched={touched}
                    httpClient={httpClient}
                    location={location}
                    notifications={notifications}
                    isDarkMode={isDarkMode}
                    flyoutMode={flyoutMode}
                  />
                </EnhancedAccordion>
                <EuiSpacer size="m" />

                <EuiSpacer size="xl" />
                <EuiTitle size="s">
                  <h3>Triggers</h3>
                </EuiTitle>
                <EuiSpacer size="m" />
                <FieldArray name={'triggerDefinitions'} validateOnChange={true}>
                  {(triggerArrayHelpers) => (
                    <ConfigureTriggers
                      triggerArrayHelpers={triggerArrayHelpers}
                      monitor={formikToMonitor(values)}
                      monitorValues={values}
                      setFlyout={setFlyout}
                      triggers={_.get(formikToMonitor(values), 'triggers', [])}
                      triggerValues={values}
                      isDarkMode={isDarkMode}
                      httpClient={httpClient}
                      notifications={notifications}
                      notificationService={notificationService}
                      plugins={plugins}
                      flyoutMode={flyoutMode}
                      submitCount={formikProps.submitCount}
                      errors={errors}
                    />
                  )}
                </FieldArray>

                <EuiSpacer size="l" />
              </div>
              <div style={{ position: 'relative' }}>
                <EuiFlexGroup justifyContent="spaceBetween" >
                  <EuiFlexItem grow={false}>
                    <EuiButton
                      onClick={() => onSubmit({ handleSubmit, validateForm })}
                      fill
                      isLoading={isLoading}
                      disabled={isDisabled}
                    >
                      {isDisabled ? 'Monitor Created'  : 'Create monitor'}
                    </EuiButton>
                  </EuiFlexItem>
                </EuiFlexGroup>
              </div>

            </div>
          );
        }}
      </Formik>
  );
}

export default AlertContainer;
