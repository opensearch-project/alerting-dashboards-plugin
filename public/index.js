/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { AlertingPlugin } from './plugin';

export function plugin(initializerContext) {
  return new AlertingPlugin(initializerContext);
}
