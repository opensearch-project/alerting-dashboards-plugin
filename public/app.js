/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { HashRouter as Router, Route } from 'react-router-dom';

import 'react-vis/dist/style.css';
// TODO: review the CSS style and migrate the necessary style to SASS, as Less is not supported in OpenSearch Dashboards "new platform" anymore
// import './less/main.less';
import './app.scss';
import Main from './pages/Main';
import { CoreContext } from './utils/CoreContext';
import { ServicesContext, NotificationService } from './services';

export function renderApp(coreStart, params) {
  const isDarkMode = coreStart.uiSettings.get('theme:darkMode') || false;
  const http = coreStart.http;
  coreStart.chrome.setBreadcrumbs([{ text: 'Alerting' }]); // Set Breadcrumbs for the plugin
  const notificationService = new NotificationService(http);
  const services = { notificationService };

  // Load Chart's dark mode CSS
  if (isDarkMode) {
    require('@elastic/charts/dist/theme_only_dark.css');
  } else {
    require('@elastic/charts/dist/theme_only_light.css');
  }

  // render react to DOM
  ReactDOM.render(
    <Router>
      <ServicesContext.Provider value={services}>
        <CoreContext.Provider
          value={{ http: coreStart.http, isDarkMode, notifications: coreStart.notifications }}
        >
          <Route render={(props) => <Main title="Alerting" {...props} />} />
        </CoreContext.Provider>
      </ServicesContext.Provider>
    </Router>,
    params.element
  );
  return () => ReactDOM.unmountComponentAtNode(params.element);
}
