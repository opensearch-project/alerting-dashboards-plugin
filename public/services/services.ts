/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { createContext } from 'react';
import { BrowserServices } from '../models/interfaces';
import { createGetterSetter } from '../../../../src/plugins/opensearch_dashboards_utils/public';
import { CoreStart, IUiSettingsClient, NotificationsStart, OverlayStart } from '../../../../src/core/public';
import { SavedObjectLoaderAugmentVis } from '../../../../src/plugins/vis_augmenter/public';
import { DataPublicPluginStart } from '../../../../src/plugins/data/public';
import { EmbeddableStart } from '../../../../src/plugins/embeddable/public';

const ServicesContext = createContext<BrowserServices | null>(null);

const ServicesConsumer = ServicesContext.Consumer;

export { ServicesContext, ServicesConsumer };

export const [getClient, setClient] =
  createGetterSetter<CoreStart['http']>('http');

export const [getSavedAugmentVisLoader, setSavedAugmentVisLoader] = createGetterSetter<
  SavedObjectLoaderAugmentVis
  >('savedAugmentVisLoader');

export const [getUISettings, setUISettings] = createGetterSetter<IUiSettingsClient>('UISettings');

export const [getEmbeddable, setEmbeddable] = createGetterSetter<EmbeddableStart>('embeddable');

export const [getOverlays, setOverlays] =
  createGetterSetter<OverlayStart>('Overlays');

export const [getQueryService, setQueryService] = createGetterSetter<
  DataPublicPluginStart['query']
  >('Query');

export const [getNotifications, setNotifications] =
  createGetterSetter<NotificationsStart>('Notifications');
