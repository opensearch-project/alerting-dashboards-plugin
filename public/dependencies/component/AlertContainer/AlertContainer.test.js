/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { httpServiceMock, notificationServiceMock, uiSettingsServiceMock} from '../../../../../../src/core/public/mocks';
import { shallow } from 'enzyme';
import { setClient, setNotifications, setUISettings} from '../../../services';
import AlertContainer from './AlertContainer';

describe('AlertContainer', () => {
  const httpClient = httpServiceMock.createStartContract();
  setClient(httpClient);
  const notifications = notificationServiceMock.createStartContract();
  setNotifications(notifications);
  const uiSettings = uiSettingsServiceMock.createStartContract();
  setUISettings(uiSettings)
  const rawContent = 'name=Flight%20Delay%20Alert&index=opensearch_dashboards_sample_data_flights&timeField=timestamp&bucketValue=12&bucketUnitOfTime=h&filters=%5B%7B%22fieldName%22%3A%5B%7B%22label%22%3A%22FlightDelayMin%22%2C%22type%22%3A%22integer%22%7D%5D%2C%22fieldValue%22%3A0%2C%22operator%22%3A%22is_greater%22%7D%5D&aggregations=%5B%7B%22aggregationType%22%3A%22sum%22%2C%22fieldName%22%3A%22FlightDelayMin%22%7D%5D&triggers=%5B%7B%22name%22%3A%22Delayed%20Time%20Exceeds%201000%20Minutes%22%2C%22severity%22%3A2%2C%22thresholdValue%22%3A1000%2C%22thresholdEnum%22%3A%22ABOVE%22%7D%5D'
  test('renders', () => {
    const wrapper = shallow(<AlertContainer content={rawContent}/>);
    print(wrapper.toString())
    expect(wrapper).toMatchSnapshot();
  });
});
