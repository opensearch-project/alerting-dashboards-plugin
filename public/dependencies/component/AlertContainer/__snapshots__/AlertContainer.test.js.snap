// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertContainer renders 1`] = `
<Formik
  initialValues={
    Object {
      "adResultIndex": undefined,
      "aggregationType": "count",
      "aggregations": Array [
        Object {
          "aggregationType": "sum",
          "fieldName": "FlightDelayMin",
        },
      ],
      "associatedMonitors": Object {
        "sequence": Object {
          "delegates": Array [],
        },
      },
      "associatedMonitorsEditor": "",
      "associatedMonitorsList": Array [],
      "bucketUnitOfTime": "h",
      "bucketValue": 12,
      "clusterNames": Array [],
      "cronExpression": "0 */1 * * *",
      "daily": 0,
      "description": "",
      "detectorId": "",
      "disabled": false,
      "fieldName": Array [],
      "filters": Array [
        Object {
          "fieldName": Array [
            Object {
              "label": "FlightDelayMin",
              "type": "integer",
            },
          ],
          "fieldValue": 0,
          "operator": "is_greater",
        },
      ],
      "frequency": "interval",
      "groupBy": Array [],
      "groupByField": Array [
        Object {
          "label": "",
        },
      ],
      "groupedOverFieldName": "bytes",
      "groupedOverTop": 5,
      "index": Array [
        Object {
          "label": "opensearch_dashboards_sample_data_flights",
        },
      ],
      "monitor_type": "query_level_monitor",
      "monthly": Object {
        "day": 1,
        "type": "day",
      },
      "name": "Flight Delay Alert-Monitor",
      "overDocuments": "all documents",
      "period": Object {
        "interval": 1,
        "unit": "MINUTES",
      },
      "preventVisualEditor": false,
      "queries": Array [],
      "query": "{
    \\"size\\": 0,
    \\"query\\": {
        \\"match_all\\": {}
    }
}",
      "searchType": "graph",
      "timeField": "timestamp",
      "timezone": Array [],
      "triggerDefinitions": Array [
        Object {
          "name": "Delayed Time Exceeds 1000 Minutes",
          "severity": 2,
          "thresholdEnum": "ABOVE",
          "thresholdValue": 1000,
        },
      ],
      "uri": Object {
        "api_type": "",
        "clusters": Array [],
        "path": "",
        "path_params": "",
        "url": "",
      },
      "weekly": Object {
        "fri": false,
        "mon": false,
        "sat": false,
        "sun": false,
        "thur": false,
        "tue": false,
        "wed": false,
      },
    }
  }
  onSubmit={[Function]}
  validateOnChange={false}
>
  <Component />
</Formik>
`;
